pkg/lib/authn/authenticator/password/provider.go

```diff
@@ -156,12 +158,34 @@ func (p *Provider) Authenticate(a *authenticator.Password, password string) (ver
 	return
 }
 
-func (p *Provider) UpdatePassword(a *authenticator.Password) error {
+// func (p *Provider) UpdatePassword(a *authenticator.Password) error {
+// 	now := p.Clock.NowUTC()
+// 	a.UpdatedAt = now
+
+// 	return nil
+// }
+
+type UpdatePasswordOptions struct {
+	PlainPassword string
+	PasswordHash  []byte
+
+	ExpireAfter    *time.Time
+	SetExpireAfter bool
+}
+
+func (p *Provider) UpdatePassword(a *authenticator.Password, options *UpdatePasswordOptions) error {
```

Comment:
I think we messed up here.

Provider.UpdatePassword should be simple. It should only do this

```
now := p.Clock.NowUTC()
a.Updated = now
err := p.Store.UpdatePasswordHash(a)
```

- Rename it to `Update`. It means Update really only updates the authenticator as is. No extra logic.

Provider.WithPassword (or Provider.WithSpec, as you have renamed it) is the real method we want to refactor. We want to support the use cases of different  password update.